include/xrpl/beast/utility/instrumentation.h

```diff
@@ -0,0 +1,58 @@
+//------------------------------------------------------------------------------
+/*
+This file is part of rippled: https://github.com/ripple/rippled
+Copyright (c) 2024 Ripple Labs Inc.
+
+Permission to use, copy, modify, and/or distribute this software for any
+purpose  with  or without fee is hereby granted, provided that the above
+copyright notice and this permission notice appear in all copies.
+
+THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef BEAST_UTILITY_INSTRUMENTATION_H_INCLUDED
+#define BEAST_UTILITY_INSTRUMENTATION_H_INCLUDED
+
+#include <cassert>
+
+#ifdef ENABLE_VOIDSTAR
+#ifdef NDEBUG
+#error "Antithesis instrumentation requires Debug build"
+#endif
+#include <antithesis_sdk.h>
+#else
+#define ALWAYS(cond, name, ...) assert((name) && (cond))
+#define ALWAYS_OR_UNREACHABLE(cond, name, ...) assert((name) && (cond))
+#define SOMETIMES(cond, name, ...)
+#define REACHABLE(name, ...)
+#define UNREACHABLE(name, ...) assert((name) && false)
+#endif
+
+#define ASSERT ALWAYS_OR_UNREACHABLE
```

Comment:
You can avoid guarding the `#include <antithesis_sdk.h>` as it already contains these define statements if their SDK is not enabled.

You could do something like:

```
#ifndef ENABLE_VOIDSTAR
#define NO_ANTITHESIS_SDK
#endif
#include <antithesis_sdk.h>
```

I left off the `NDEBUG` check above, because the cmake config already takes care of when the voidstar option can and cannot be turned on. And the Antithesis header then has the following check to determine whether or not to enable the SDK: `#if defined(NO_ANTITHESIS_SDK) || __cplusplus < 202000L || (defined(__clang__) && __clang_major__ < 16)`.

That being said, the Antithesis header does have a few system library includes plus the namespace definition containing a struct (top of file) and template function (bottom of file). If it is indeed a concern that these would be included in all builds, including releases, then your current code would be fine. That being said, if you're open to modifying the `antithesis_sdk.h` file then you could possibly move that `#if defined(NO_ANTITHESIS_SDK) ...` statement around the namespace or at least the struct/function to get the best of both worlds.

Reply:
The idea was to entirely exclude the dependency on `external/antithesis-sdk` if the build is not using the `voidstar` option. The reason is that, at least for the time being, we want `rippled` to build successfully on a wider selection of compilers than supported by `external/antithesis-sdk` - which includes clang 15, which would trip on this:

```
#if defined(__clang__) && __clang_major__ < 16
    #error "The Antithesis C++ API requires clang version 16 or higher"
#endif
```

Having said that I like the idea of not having to repeat definitions which are in this section of `#include <antithesis_sdk.h>` ; maybe I can talk Antithesis into doing something to make this header easier to swallow for clang 15 (when `NO_ANTITHESIS_SDK` is defined)

Reply:
No wait, I would still need to guard `#include <antithesis_sdk.h>`, because (if `voidstar` is not being used) I need to define all the polyfills for instrumentation macros. The polyfills from `<antithesis_sdk.h>` are not good, as they do not `assert`.

Reply:
Ah yes, you are correct - they just define the names but don't attach any actions to them.